Event schema

d-chambers opened the pull request:
This PR adds an event schema using [pydantic](https://pydantic-docs.helpmanual.io/) then uses pydantic to to perform the conversions of obspy.Catalog objects to and from json/dicts.

This is primarily an internal change and shouldn't really effect ObsPlus' external API.

d-chambers commented:
I should mention I had to slightly modify test catalog 4 (tests/test_data/test_catalogs/cat4.xml) because ObsPy's quakeml parser expected ints for `CompositeTime`'s year, month, day, hour, and minute uncertainties. However, this is inconsistent with `QuantityError`specifying its types as floats. Since this is such an obscure corner case (I have never seen a `CompositeTime` in the wild) I just removed the uncertainties from the test catalog.

d-chambers commented:
I ended up finding a bug where `WaveBank` couldn't be pickled because the index cached used itertools.cycle. I replaced this with a simpler incrementing index attribute.

d-chambers commented:
Anyone have a sec to review this? I am not too worried about it since the external APIs haven't changed but another pair of eyes couldn't hurt.

calum-chamberlain commented on obsplus/events/json.py:
```diff
@@ -6,136 +6,44 @@
 
 
-def cat_to_json(catalog: obspy.Catalog) -> str:
+def cat_to_json(events: Union[Catalog, Event]) -> str:
```
Good to see single events explicitly allowed, but would it be possible to keep the variable as `catalog`? I'm pretty sure I use this function directly in a few of my codes, and I'm concerned that I may have used a kwarg at some point.

calum-chamberlain commented on obsplus/events/json.py:
```diff
@@ -6,136 +6,44 @@
 
 
-def cat_to_dict(obj):
+def cat_to_dict(events: Union[Catalog, Event]) -> dict:
```
Nice change in this function - I'm pretty sure I wouldn't have used an `obj` kwarg, but if you go with the `catalog` kwarg above, can you use `catalog` here as well?

calum-chamberlain commented on obsplus/events/schema.py:
```diff
@@ -0,0 +1,612 @@
+"""
```
I'm going to assume this is all there given the tests. I haven't checked through this fully though!

d-chambers replied:
Yes, there is at least one test event (#4) which uses every kind of attribute supported by ObsPy. There is a little edge case where the `CompositeTime`'s uncertainties corresponding to int fields don't work (as mentioned above). This is really more of a bug in ObsPy's parsing/typing though and I don't think that it will really effect anyone.

calum-chamberlain commented on tests/test_events/test_json.py:
```diff
@@ -73,6 +73,22 @@ def test_catalog_can_be_written(self, test_catalog, json_cat_from_disk):
+    def test_list_of_events(self):
+        """Ensure a list of events can be converted to json."""
+        events = obspy.read_events().events
+        out = cat_to_json(events)
```
Should the docstring for `cat_to_json` should be `Union[Catalog, Event, Iterable[Event]]`?

calum-chamberlain reviewed (changes requested):
Looks good to me - I had a couple of pedantic comments for this pydantic proposal.

d-chambers commented:
Thanks @calum-chamberlain, I think I addressed all your concerns, thanks for taking a look at it.

I also added a changelog entry and restructured `cat_to_dict` to use the pydantic model's `dict` method rather than first converting to json which resulted in a minor speedup.

d-chambers commented:
I also should document here that I timed the json test suite before and after this change. Before using pydantic the test suite took on average 14.5 seconds. After using pydantic the test suite takes about 15 seconds. Considering the huge improvement to the code readability and elimination of recursive functions I think the minor slow-down is worth it.
